### ISSUE TITLE
🐛 Fatal exception: CameraDevice was already closed  

### ISSUE THREAD

**Issue by fsrajner** [cuvent/react-native-vision-camera#2560]
### What's happening?

Hi! I'm trying to use the camera simply to take and save pictures, but after using it for more than some test photos, the app crashes.

Behavior:
Opening and closing the camera repeatedly, and focusing randomly causes the application to crash within a minute.

Expected Behavior:
Not crashing, or at worst restarting the camera device.

### Reproduceable Code

```tsx
import * as React from 'react'
import { useRef, useState, useCallback, useMemo, useEffect } from 'react'
import { AppState, AppStateStatus, Pressable, StyleSheet, View } from 'react-native'
import { Gesture, GestureDetector } from 'react-native-gesture-handler'
import {
    CameraDevice, CameraRuntimeError, Code, Orientation, PhotoFile, Point, useCameraDevice,
    useCameraDevices, useCameraFormat, useCameraPermission, useCodeScanner, VideoFile, Camera as VisionCamera,

    CameraProps as VisionCameraProps
} from 'react-native-vision-camera'
import Reanimated, { Easing, Extrapolation, interpolate, runOnJS, useAnimatedProps, useAnimatedStyle, useSharedValue, withRepeat, withSpring, withTiming } from 'react-native-reanimated'
import { useIsFocused } from '@react-navigation/core'
import { Icon } from '../Icon'
import { colors, layout, spacing } from 'app/theme'
import { Button } from '../Button'
import { alert, orientationCalculation } from 'app/utils/utils'
import { translate } from 'app/i18n'
import { DeviceMotion } from 'expo-sensors';

Reanimated.addWhitelistedNativeProps({
    zoom: true,
})
const ReanimatedCamera = Reanimated.createAnimatedComponent(VisionCamera)

const useIsForeground = (): boolean => {
    const [isForeground, setIsForeground] = useState(true)

    useEffect(() => {
        const onChange = (state: AppStateStatus): void => {
            setIsForeground(state === 'active')
        }
        const listener = AppState.addEventListener('change', onChange)
        return () => listener.remove()
    }, [setIsForeground])

    return isForeground
}

function usePreferredCameraDevice(_preferredDeviceId: string): [CameraDevice | undefined, (device: CameraDevice) => void] {
    const [preferredDeviceId, setPreferredDeviceId] = useState(_preferredDeviceId)

    const set = useCallback(
        (device: CameraDevice) => {
            setPreferredDeviceId(device.id)
        },
        [setPreferredDeviceId],
    )

    const devices = useCameraDevices()
    const device = useMemo(() => devices.find((d) => d.id === preferredDeviceId), [devices, preferredDeviceId])

    return [device, set]
}

export interface CameraProps {
    /**
     * An optional style override useful for padding & margin.
     */
    // style?: StyleProp<ViewStyle>

    onMediaCaptured?: (media: PhotoFile | VideoFile, type: 'photo' | 'video') => void

    onCodeScanned?: (codes: Code[]) => void,

    showControls?: boolean
}

export function Camera(props: CameraProps) {
    const { onMediaCaptured, onCodeScanned, showControls } = props

    const { hasPermission, requestPermission } = useCameraPermission()


    const requestCameraPermission = useCallback(async () => {
        console.log('Requesting camera permission...')
        if (!hasPermission)
            console.log(`Camera permission status: ${requestPermission()}`)
    }, [])

    const camera = useRef<VisionCamera>(null)

    // check if camera page is active
    const [isCameraInitialized, setIsCameraInitialized] = useState(false)

    const isFocused = useIsFocused()
    const isForeground = useIsForeground()
    const isActive = isFocused && isForeground

    const [cameraPosition, setCameraPosition] = useState<'front' | 'back'>('back')
    const [flash, setFlash] = useState<'off' | 'on'>('off')
    const [aspectRatio, setAspectRatio] = useState({
        width: 4,
        height: 3
    })

    // camera device settings
    const [preferredDevice] = usePreferredCameraDevice(cameraPosition)
    let device = useCameraDevice(cameraPosition)

    if (preferredDevice != null && preferredDevice.position === cameraPosition) {
        // override default device with the one selected by the user in settings
        device = preferredDevice
    }
    const format = useCameraFormat(device, [
        // { photoAspectRatio: aspectRatio },
        { photoResolution: 'max' },
        { photoAspectRatio: aspectRatio.width / aspectRatio.height }
    ])

    const supportsFlash = device?.hasFlash ?? false
    useEffect(() => {
        if (device && isCameraInitialized) {
            const formats = device.formats.filter(x => typeof x.photoHeight === "number" && typeof x.photoWidth === "number")
            if (formats.length > 0) {
                const best = formats.reduce((acc, current) => {
                    if (current.photoHeight * current.photoWidth > acc.photoHeight * acc.photoWidth)
                        return current
                    else
                        return acc
                }, formats[0])

                setAspectRatio({
                    width: best.photoWidth,
                    height: best.photoHeight
                })
            } else {
                setAspectRatio({
                    width: 4,
                    height: 3
                })
            }
        }
    }, [device, isCameraInitialized])

    // Camera callbacks
    const onInitialized = useCallback(() => {
        console.log('Camera initialized!')
        setIsCameraInitialized(true)
    }, [])

    const onError = useCallback((error: CameraRuntimeError) => {
        console.error(error)
    }, [])

    const onFlipCameraPressed = useCallback(() => {
        setCameraPosition((p) => (p === 'back' ? 'front' : 'back'))
    }, [])
    const onFlashPressed = useCallback(() => {
        setFlash((f) => (f === 'off' ? 'on' : 'off'))
    }, [])
    // #endregion

    const [inFocus, setInFocus] = useState(false)
    const [focusPoint, setFocusPoint] = useState({ x: -1, y: -1 })
    const focus = useCallback((point: Point) => {
        const c = camera.current
        if (c == null) return
        setInFocus(true)
        setFocusPoint(point)
        setTimeout(() => {
            setInFocus(false)
        }, 3000)
        c.focus(point).then(() => setInFocus(false)).catch((error) => { console.log(error) })
    }, [])
    const focusGesture = Gesture.Tap()
        .onEnd(({ x, y }) => {
            runOnJS(focus)({ x, y })
        })

    const _onCodeScanned = useCallback((codes: Code[]) => {
        console.log(`Scanned ${codes.length} codes:`, codes)
        const value = codes[0]?.value
        if (value == null || !onCodeScanned) return
        onCodeScanned(codes)
    }, [])
    const codeScanner = onCodeScanned ? useCodeScanner({
        codeTypes: ['qr', 'ean-13', 'codabar', 'code-128', 'aztec', 'code-93', 'data-matrix', 'ean-8', 'itf', 'upc-e'],
        onCodeScanned: _onCodeScanned,
    }) : null


    useEffect(() => {
        // const f =
        //     format != null
        //         ? `(${format.photoWidth}x${format.photoHeight} photo / ${format.videoWidth}x${format.videoHeight}@${format.maxFps})`
        //         : undefined
        // console.log(`Camera: ${JSON.stringifydevice?.name} | Format: ${f}`)
        console.log(`| Camera: ${JSON.stringify(device, (k, v) => k === "formats" ? [] : v, 2)}`)
    }, [device?.name, format])


    useEffect(() => {
        if (!isActive)
            device = null
    }, [isActive])


    //#region zoom

    const zoom = useSharedValue(device.neutralZoom)

    const zoomOffset = useSharedValue(0);
    const zoomGesture = Gesture.Pinch()
        .onBegin(() => {
            zoomOffset.value = zoom.value
        })
        .onUpdate(event => {
            const z = zoomOffset.value * event.scale
            zoom.value = interpolate(
                z,
                [1, 10],
                [device.minZoom, device.maxZoom],
                Extrapolation.CLAMP,
            )
        })

    const animatedProps = useAnimatedProps<VisionCameraProps>(
        () => ({ zoom: zoom.value }),
        [zoom]
    )
    //#endregion
    //#region handle orientation
    const [orientation, setOrientation] = useState('portrait' as Orientation);
    const [subscription, setSubscription] = useState(null);

    const _subscribe = async () => {
        const available = await DeviceMotion.isAvailableAsync()
        if (available) {
            let permission = await DeviceMotion.getPermissionsAsync()
            if (!permission.granted) {
                permission = await DeviceMotion.requestPermissionsAsync()
            }
            if (permission.granted)
                setSubscription(
                    DeviceMotion.addListener(data => {
                        let cameraOrientation = 'portrait' as Orientation
                        try {
                            const calculatedOrientation = orientationCalculation(data.rotation.gamma, data.rotation.beta)
                            switch (calculatedOrientation) {
                                case 0:
                                    cameraOrientation = 'portrait'
                                    break;
                                case 90:
                                    // cameraOrientation = 'landscape-right'
                                    cameraOrientation = 'landscape-left'
                                    break;
                                case 180:
                                    cameraOrientation = 'portrait-upside-down'
                                    break;
                                case -90:
                                    cameraOrientation = 'landscape-right'
                                    // cameraOrientation = 'landscape-left'
                                    break;
                            }
                        } catch (error) {

                        }
                        setOrientation(cameraOrientation);
                    })
                );
        }
    };

    const _unsubscribe = () => {
        subscription && subscription.remove();
        setSubscription(null);
    };

    useEffect(() => {
        DeviceMotion.setUpdateInterval(200)
        _subscribe();
        return () => _unsubscribe();
    }, []);
    //#endregion


    //region capture button
    const [enabled, setEnabled] = useState(true)
    const [isPressingButton, setIsPressingButton] = useState(false)
    const CAPTURE_BUTTON_SIZE = spacing.xxxl
    const BORDER_WIDTH = CAPTURE_BUTTON_SIZE * 0.1
    const shadowStyle = useAnimatedStyle(
        () => ({
            transform: [
                {
                    scale: withSpring(isPressingButton ? 1 : 0, {
                        mass: 1,
                        damping: 35,
                        stiffness: 300,
                    }),
                },
            ],
        }),
        [isPressingButton],
    )
    const buttonStyle = useAnimatedStyle(() => {
        let scale: number
        if (enabled) {
            if (isPressingButton) {
                scale = withRepeat(
                    withSpring(1, {
                        stiffness: 100,
                        damping: 1000,
                    }),
                    -1,
                    true,
                )
            } else {
                scale = withSpring(0.9, {
                    stiffness: 500,
                    damping: 300,
                })
            }
        } else {
            scale = withSpring(0.6, {
                stiffness: 500,
                damping: 300,
            })
        }

        return {
            opacity: withTiming(enabled ? 1 : 0.3, {
                duration: 10,
                easing: Easing.linear,
            }),
            transform: [
                {
                    scale,
                },
            ],
        }
    }, [enabled, isPressingButton])
    //#endregion
    if (!isActive || !hasPermission || device == null)
        return (
            <View style={styles.container}>
                <View style={[layout.fillHeight, layout.center]}>
                    <Button
                        style={layout.centerButton}
                        textStyle={layout.centerButtonText}
                        tx={"dialogs.enableCamera"}
                        onPress={() => requestCameraPermission()}
                    />
                </View>
            </View>)
    else
        return (
            <View style={styles.container}>
                <View style={{
                    width: "100%",
                    height: "auto",
                    aspectRatio: `${aspectRatio.height} / ${aspectRatio.width}`,
                    backgroundColor: "black"
                }}>

                    <GestureDetector gesture={zoomGesture}>
                        <GestureDetector gesture={focusGesture}>
                            <ReanimatedCamera
                                ref={camera as any}
                                style={{
                                    width: "100%",
                                    height: "100%",
                                }}
                                // style={StyleSheet.absoluteFill}
                                device={device}
                                format={format}
                                lowLightBoost={true}
                                isActive={isActive}
                                onInitialized={onInitialized}
                                onError={onError}
                                enableZoomGesture={false}
                                animatedProps={animatedProps}
                                // exposure={0}
                                orientation={orientation}
                                photo={true}
                                video={false}
                                codeScanner={onCodeScanned ? codeScanner : undefined}
                            />
                        </GestureDetector>
                    </GestureDetector>
                </View>
                {
                    showControls !== false &&
                    <View style={[layout.aboveStyle,
                    {
                        width: "100%",
                        height: "100%",
                        display: "flex"
                    }]}>
                        {inFocus && <View style={{
                            height: 48,
                            width: 48,
                            borderColor: colors.magenta,
                            borderWidth: spacing.xxxs,
                            borderRadius: 12,
                            backgroundColor: 'transparent',
                            position: 'absolute',
                            zIndex: 9005,
                            elevation: 9005,
                            top: focusPoint.y - 24,
                            left: focusPoint.x - 24
                        }} />}
                        <View
                            style={[{
                                width: "100%",
                                flexDirection: "row",
                                alignItems: "flex-end",
                                alignContent: "flex-end",
                                justifyContent: "flex-end"
                            }]}>
                            {supportsFlash && (
                                <Icon icon={flash === 'on' ? 'flash' : 'flashOff'} color="white" size={spacing.xl} onPress={onFlashPressed} style={{ marginHorizontal: spacing.xs }} />
                            )}
                            <Icon icon="cameraReverse" color="white" size={spacing.xl} onPress={onFlipCameraPressed} style={{ marginHorizontal: spacing.xs }} />

                        </View>
                        {/* <View style={{
                            width: "100%",
                            height: "auto",
                            aspectRatio: `${aspectRatio.height} / ${aspectRatio.width}`,
                        }} >
                        </View> */}

                        <View style={[{
                            width: "100%",
                            justifyContent: "flex-end",
                            flex: 1
                        }]}>
                            <View style={{
                                width: "100%",
                                paddingBottom: spacing.xl,
                                justifyContent: "center",
                                alignContent: "center",
                                alignItems: "center"
                            }}>
                                <Reanimated.View style={buttonStyle}>
                                    <Reanimated.View style={[
                                        {
                                            backgroundColor: colors.magenta,
                                            borderRadius: CAPTURE_BUTTON_SIZE / 2,
                                            height: CAPTURE_BUTTON_SIZE,
                                            position: 'absolute',
                                            width: CAPTURE_BUTTON_SIZE,
                                        },
                                        , shadowStyle]} />
                                    <Pressable style={{
                                        borderColor: 'white',
                                        borderRadius: CAPTURE_BUTTON_SIZE / 2,
                                        borderWidth: BORDER_WIDTH,
                                        height: CAPTURE_BUTTON_SIZE,
                                        width: CAPTURE_BUTTON_SIZE,
                                    }}
                                        onPressIn={() => setIsPressingButton(true)}
                                        onPressOut={async () => {
                                            try {
                                                setIsPressingButton(false)
                                                setEnabled(false)
                                                const photo = await camera.current?.takePhoto({
                                                    flash
                                                })
                                                onMediaCaptured(photo, "photo")
                                            } catch (error) {
                                                alert.alert({
                                                    header: translate("alert.error"),
                                                    text: error
                                                })
                                            } finally {
                                                setEnabled(true)
                                            }

                                        }}
                                    />
                                </Reanimated.View>
                            </View>
                        </View>
                    </View>
                }
            </View >
        )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        display: "flex"
    },
})
```

### Relevant log output

```shell
02-14 16:19:39.114   798  1210 I ExynosCameraMetadataConverter: [CAM(0)][]-(translateRequestControlData[2502]):ANDROID_REQUEST_ID(2)
02-14 16:19:39.114   798  3756 I EDEN    : [Exynos][EDEN][v1.6.20][RT::EdenRuntime] ExecuteReq:353: (+) modelId=79804123, mode=0 req=0x7a1e3517b0
02-14 16:19:39.121   798  3766 I ExynosCameraMetadataConverterVendor: [CAM(0)][]-(translateVendorControlMetaData[4432]):[R14 F16] dynamicShotCondition (780004 -> 790003)
02-14 16:19:39.150   798  1210 I ExynosCameraMetadataConverter: [CAM(0)][]-(translateRequestControlData[2502]):ANDROID_REQUEST_ID(3)
02-14 16:19:39.179  3207  3248 I PersistentCameraCaptureSession: Resetting focus to auto-focus...
02-14 16:19:39.183  3207  3248 E AndroidRuntime: FATAL EXCEPTION: mrousavy/VisionCamera.main
02-14 16:19:39.183  3207  3248 E AndroidRuntime: Process: com.infom.kingmobile.internal, PID: 3207
02-14 16:19:39.183  3207  3248 E AndroidRuntime: java.lang.IllegalStateException: CameraDevice was already closed
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.hardware.camera2.impl.CameraDeviceImpl.checkIfCameraClosedOrInError(CameraDeviceImpl.java:2456)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.hardware.camera2.impl.CameraDeviceImpl.createCaptureRequest(CameraDeviceImpl.java:863)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at com.mrousavy.camera.core.capture.CameraCaptureRequest.createCaptureRequest(CameraCaptureRequest.kt:52)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at com.mrousavy.camera.core.capture.RepeatingCaptureRequest.createCaptureRequest(RepeatingCaptureRequest.kt:49)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at com.mrousavy.camera.core.capture.RepeatingCaptureRequest.createCaptureRequest(RepeatingCaptureRequest.kt:33)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at com.mrousavy.camera.core.PersistentCameraCaptureSession$focus$2$4.invokeSuspend(PersistentCameraCaptureSession.kt:243)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:235)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:168)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:474)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:508)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:497)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:595)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at kotlinx.coroutines.android.HandlerContext$scheduleResumeAfterDelay$$inlined$Runnable$1.run(Runnable.kt:19)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:942)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:226)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:313)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at android.os.HandlerThread.run(HandlerThread.java:67)
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@2887bc8, mrousavy/VisionCamera.main]
02-14 16:19:39.190  1164  3788 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
```

### Camera Device

```json
{
    "formats": [],
    "sensorOrientation": "landscape-left",
    "hardwareLevel": "full",
    "maxZoom": 8,
    "minZoom": 1,
    "maxExposure": 20,
    "supportsLowLightBoost": true,
    "neutralZoom": 1,
    "physicalDevices": [
        "wide-angle-camera"
    ],
    "supportsFocus": true,
    "supportsRawCapture": true,
    "isMultiCam": false,
    "minFocusDistance": 10,
    "minExposure": -20,
    "name": "BACK (0)",
    "hasFlash": true,
    "hasTorch": true,
    "position": "back",
    "id": "0"
}
```

### Device

Samsung Galaxy S20 FE

### VisionCamera Version

3.9.0-beta.3

### Can you reproduce this issue in the VisionCamera Example app?

I didn't try (⚠️ your issue might get ignored & closed if you don't try this)

### Additional information

- [ ] I am using Expo
- [ ] I have enabled Frame Processors (react-native-worklets-core)
- [X] I have read the [Troubleshooting Guide](https://react-native-vision-camera.com/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

**Comment by mrousavy** [comment]
This is in your logs:

```
02-14 16:19:39.183  3207  3248 E AndroidRuntime: 	at com.mrousavy.camera.core.PersistentCameraCaptureSession$focus$2$4.invokeSuspend(PersistentCameraCaptureSession.kt:243)
```

I'm guessing you are calling `focus()`, right?

**Comment by mrousavy** [comment]
...if yes, then I already fixed this in this PR: https://github.com/mrousavy/react-native-vision-camera/pull/2550

Will release a new version later

**Comment by fsrajner** [comment]
Yes, I'm just copying the example code. Thanks for the response. will update as soon as it releases :)

**Comment by mrousavy** [comment]
Hey - I just released [VisionCamera 3.9.0-beta.4](https://github.com/mrousavy/react-native-vision-camera/releases/tag/v3.9.0-beta.4) which probably fixes this issue - can you try that latest beta and let me know? :)

**Comment by fsrajner** [comment]
Thanks, it works!
